### PR TITLE
SoC2023 RFSB509

### PR DESCRIPTION
**Pull request by MxAR**
# About
In the context of this years Summer of Code I have implemented the RFSB509 cryptographic hash function [1] in this branch. This function depends on the SHA256 algorithm and AES128 cipher procedure. A SHA256 implementation already exists in this library so this branch only adds the AES128 (Inv_)Cipher procedure. Additionally, because I initially misread the RFSB509 paper I also added the AES256 (Inv_)Cipher procedure.  The AES procedures were implemented according to the NIST FIPS 197 standard and optimized using information from [2]. All subprograms are implemented to avoid timing attacks, this  includes no integer multiplications, no table lookups and no branching based on secret data. 

## References
1) Bernstein, D.J., Lange, T., Peters, C., Schwabe, P. (2011). Really Fast
    Syndrome-Based Hashing. In: Nitaj, A., Pointcheval, D. (eds) Progress in
    Cryptology – AFRICACRYPT 2011. AFRICACRYPT 2011. Lecture Notes in
    Computer Science, vol 6737. Springer, Berlin, Heidelberg.
    https://doi.org/10.1007/978-3-642-21969-6_9
2) Satoh, A., Morioka, S., Takano, K., Munetoh, S. (2001).  A Compact
    Rijndael Hardware Architecture with S-Box Optimization.  In: Boyd, C.
    (eds) Advances in Cryptology — ASIACRYPT 2001. ASIACRYPT 2001.  Lecture
    Notes in Computer Science, vol 2248. Springer, Berlin, Heidelberg.
    https://doi.org/10.1007/3-540-45682-1_15

### CODE REVIEW

**Comment by rod-chapman** [comment]
We'll squash after review.

**Comment by rod-chapman** [comment]
Still fails on both M1 and Intel macOS with

```
   [link]         testall.adb
Undefined symbols for architecture arm64:
  "_rfsb509_crypto_hash", referenced from:
      __ada_rfsb509_supercop_regression in rfsb509_supercop_regression.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
gprbuild: link of testall.adb failed
```

**Comment by rod-chapman** [comment]
Build of "tests" now looks good on both M1 and Intel macOS

**Comment by docandrew** [comment]
I too was able to get a successful build and passing tests on M1 w/ GNAT 12.2.1 and gprbuild 22.0.1.

**Comment by rod-chapman** [comment]
Which version of the RISC-V cross compiler are you using to compile the performance tests? I think I need an upgrade... mine is very old...

**Comment by rod-chapman** [comment]
Hmmm.. I installed the risc64-elf 12.2.1 toolchain (using alr 1.2.2) on my Linux VM.
Now in perf//aes256 I get:

```
rchapman@buster:~/prog/maxnacl/perf/aes256$ make
gprbuild -j4 -Paes256 -v -XSPARKNACL_RUNTIME_MODE=zfp -XSPARKNACL_RUNTIME_CHECKS=disabled -XSPARKNACL_CONTRACTS=disabled -XSPARKNACL_BUILD_MODE=O2 -XSPARKNACL_TARGET_ARCH=rv32imc_a4 
gprconfig: can't find a toolchain for the following configuration:
gprconfig: language 'ada', target 'riscv64-elf', runtime 'zfp-rv32imac'

==============Messages for file: /home/rchapman/prog/maxnacl/sparknacl.gpr
     1. library project Sparknacl is
                        |
        >>> no compiler for language "Ada", cannot compile "sparknacl-mac.adb"
gprbuild: *** compilation phase failed
make: *** [makefile:12: obj/aes256] Error 4
```

**Comment by rod-chapman** [comment]
Yes.. PATH looks like this:
```
/home/rchapman/gnat/alr1.2.2/bin:
/home/rchapman/.config/alire/cache/dependencies/gprbuild_22.0.1_24dfc1b5/bin:
/home/rchapman/gnat/spark-pro-23.0w-20220718-x86_64-linux-bin/bin:
/home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin:
/home/rchapman/.config/alire/cache/dependencies/gnat_native_12.1.1_df1f36b3/bin:
/media/psf/Home/Desktop/prog/AdaCore/gnatstack-20.2-20200603-1DACB-src/bin
```

**Comment by rod-chapman** [comment]
```
rchapman@buster:~/prog/maxnacl/perf/aes256$ gprconfig --target=riscv64-elf
--------------------------------------------------
gprconfig has found the following compilers on your PATH.
Only those matching the target and the selected compilers are displayed.
   1. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (embedded-polarfiresoc runtime)
   2. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-hifive1 runtime)
   3. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-polarfiresoc runtime)
   4. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32i runtime)
   5. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32iac runtime)
   6. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32im runtime)
   7. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32imac runtime)
   8. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32imafc runtime)
   9. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv32imafdc runtime)
  10. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64im runtime)
  11. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64imac runtime)
  12. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64imafc runtime)
  13. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64imafdc runtime)
  14. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64imc runtime)
  15. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-rv64imfc runtime)
  16. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-spike runtime)
  17. GNAT for Ada in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2 (light-tasking-polarfiresoc runtime)
  18. GCC-ASM for Asm in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2.0
  19. GCC-ASM for Asm2 in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2.0
  20. GCC-ASM for Asm_Cpp in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2.0
  21. LD for Bin_Img in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 2.39
  22. GCC for C in /home/rchapman/.config/alire/cache/dependencies/gnat_riscv64_elf_12.2.1_fc4a4d08/bin/ version 12.2.0
Select or unselect the following compiler (or "s" to save): ^C
r
```

**Comment by rod-chapman** [comment]
Ah... I think I need to update Ada_Drivers_Library so it specified the "light" runtime rather than "zip"...

**Comment by rod-chapman** [comment]
How come SPARKNaCl.Hashing.RFSB509.Hash takes a "Key" formal parameter, but SUPERCOP_RFSB509_API.SUPERCOP_RFSB509_Hash does not? This needs to be explained...

**Comment by rod-chapman** [comment]
Both please!

**Comment by rod-chapman** [comment]
OK... Understood. I think a comments in the package spec to say "this is the keyed variant", and "Use (others => 0) as the key to get same behaviour as the SUPERCOP reference code" should suffice.  That also explains why that value for Key is used in the test program...

File: tests/src/ada/aes256_cipher_kat.adb [hunk 1]
```diff
@@ -0,0 +1,158 @@
+with SPARKNaCl;     use SPARKNaCl;
+with SPARKNaCl.AES; use SPARKNaCl.AES;
+
+with Ada.Text_IO; use Ada.Text_IO;
+
+procedure AES256_Cipher_KAT
+is
+   Test_Case_Count : constant I32 := 12;
+
+   Keys : constant array (1 .. Test_Case_Count) of Bytes_32 := (
```
Comment (rod-chapman) [inline tests/src/ada/aes256_cipher_kat.adb:10]: Where do these KATs come from? Can you cite a reference to the source please as a comment here? Same comment for aes128_cipher_kat.adb

Comment (rod-chapman) [reply tests/src/ada/aes256_cipher_kat.adb:10]: OK... but give me the precise ref to the PDF and Section number. e.g.
"Annex B, C, and E of https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/aes/AESAVS.pdf"

as a comment.

**Comment by rod-chapman** [comment]
So add that link and explanation as a comment just above the first declaration.

**Review by rod-chapman — APPROVED** [review]
Looks all good now.

**Pull request merged**